File: dotnet/src/PqcStandards/MlDsa/DsaField.cs
namespace PqcStandards.MlDsa;

/// <summary>Field arithmetic modulo Q = 8380417 for ML-DSA.</summary>
public static class DsaField
{
    public const int Q = 8380417;

    public static int ModQ(long a)
    {
        int r = (int)(a % Q);
        return r < 0 ? r + Q : r;
    }

    public static int Add(int a, int b) => ModQ((long)a + b);
    public static int Sub(int a, int b) => ModQ((long)a - b);
    public static int Mul(int a, int b) => ModQ((long)a * b);

    /// <summary>Modular exponentiation.</summary>
    public static int Pow(int a, int exp)
    {
        long result = 1;
        long bas = ModQ(a);
        if (exp < 0) exp += Q - 1;
        while (exp > 0)
        {
            if ((exp & 1) == 1) result = (result * bas) % Q;
            bas = (bas * bas) % Q;
            exp >>= 1;
        }
        return (int)result;
    }

    /// <summary>Centered representative in [-(Q-1)/2, (Q-1)/2].</summary>
    public static int CenterMod(int a)
    {
        int r = a % Q;
        if (r < 0) r += Q;
        return r > Q / 2 ? r - Q : r;
    }
}

File: php/src/MlDsa/DsaField.php
<?php

declare(strict_types=1);

namespace PQC\MlDsa;

/**
 * Field arithmetic modulo q=8380417 for ML-DSA.
 */
final class DsaField
{
    public const Q = 8380417;
    public const Q_HALF = 4190208; // (Q-1)/2

    /**
     * Reduce mod q into [0, q).
     */
    public static function mod(int $a): int
    {
        $r = $a % self::Q;
        return $r < 0 ? $r + self::Q : $r;
    }

    /**
     * Addition mod q.
     */
    public static function add(int $a, int $b): int
    {
        return self::mod($a + $b);
    }

    /**
     * Subtraction mod q.
     */
    public static function sub(int $a, int $b): int
    {
        return self::mod($a - $b);
    }

    /**
     * Multiplication mod q.
     */
    public static function mul(int $a, int $b): int
    {
        return self::mod($a * $b);
    }

    /**
     * Power mod q.
     */
    public static function pow(int $base, int $exp): int
    {
        $result = 1;
        $base = self::mod($base);
        while ($exp > 0) {
            if ($exp & 1) {
                $result = self::mul($result, $base);
            }
            $exp >>= 1;
            $base = self::mul($base, $base);
        }
        return $result;
    }

    /**
     * Inverse mod q.
     */
    public static function inv(int $a): int
    {
        return self::pow($a, self::Q - 2);
    }

    /**
     * Centered representative: map [0, q) to [-(q-1)/2, (q-1)/2].
     */
    public static function centered(int $a): int
    {
        $a = self::mod($a);
        if ($a > self::Q_HALF) {
            return $a - self::Q;
        }
        return $a;
    }

    /**
     * Infinity norm of a polynomial.
     */
    public static function polyNorm(array $poly): int
    {
        $max = 0;
        foreach ($poly as $c) {
            $v = abs(self::centered($c));
            if ($v > $max) {
                $max = $v;
            }
        }
        return $max;
    }

    /**
     * Infinity norm of a vector of polynomials.
     */
    public static function vecNorm(array $vec): int
    {
        $max = 0;
        foreach ($vec as $poly) {
            $n = self::polyNorm($poly);
            if ($n > $max) {
                $max = $n;
            }
        }
        return $max;
    }
}
